### ISSUE TITLE
changing controller can cause OOS (GNA #25182)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1586]
**[Original submission](https://gna.org/bugs/index.php?25182)** by gfgtdf on 2016-10-14

Assume a mp scenario with 2 clients, side 1 is host, its turn of side 2. when side 1 now uses :control command to reassign side 1 to another player client 2 might get an OOS in case that he did moved at the same time (since the server will rekect those moves but the client has already done them and cannot undo them.)

Thre are multilpe was to fix the problem, for example:

1) we could make :control work in 2 phases where first everyones moves are blockes and the the sides controllers are changes, like in:

```

1) host -> server ':control'

2) server -> c 'lock moves'

3) current player -> server 'lock done'

4) server -> all clients 'change controller and release lock'

```

where step 4 only happens after the server recived 'lock done' form the current player clients.

2) Another was to fix this problem would be to make the server smart enough to notice that a package was rejected because of an async controller change and then send the curent player the correct gamestate to fix the OOS.

3) Another option, not really a fix but sureley an improvement of the current situation would be to make the server smart enough to notice that a package was rejected because of an async controller change and then send the cilent a clear message of what has happened istead of just a 'rejected package' message.

(Reproduced on all)
Release: 1.13, 1.12, and earlier
Priority: 5 - Normal
Severity: 3 - Normal

**Comment by gfgtdf** [comment]
as noted in #1301 this is quite easy to reproduce by using a modal dialog.

**Comment by soliton-** [comment]
I can only make sense of this if the host reassigns side 2.

**Comment by gfgtdf** [comment]
> I can only make sense of this if the host reassigns side 2.

Yes that's what this report is about.

**Comment by soliton-** [comment]
> Yes that's what this report is about.

It'd help if you edit your initial post to make it say that.

**Comment by Wedge009** [comment]
#1483 and #1490 also seem to discuss OoS issues related to changing controller.

**Comment by stevecotton** [comment]
> 2) server -> current player: 'lock moves'
> 3) current player -> server: 'lock done'

That would need a fairly quick timeout, on the assumption that a host doing a controller change would generally be because the player appears disconnected or AFK.

**Comment by gfgtdf** [comment]
> That would need a fairly quick timeout, on the assumption that a host doing a controller change would generally be because the player appears disconnected or AFK.


What exactly would "quick" here mean for you ?

**Comment by stevecotton** [comment]
"Quick" would need to be under a second, although some UI might extend that a bit.

The reason is that the host is already thinking "either that player is disconnected, or I'm disconnected", and the host is taking action to fix that. So the response needs to happen before the host decides that they're the one that's disconnected and quits.

**Comment by gfgtdf** [comment]
> The reason is that the host is already thinking "either that player is disconnected, or I'm disconnected",

I don't believe change_controller should be (or is) the  first reaction in such a case. I would for example do a less invasive server query to check whether i'm disconnected or not. Also we could give the host a immediate  response even if it takes the other client a little longer to react.
